### PR TITLE
Type checking for authorized references

### PR DESCRIPTION
**Pull request by dsainati1**
Part of the implementation of https://github.com/onflow/flips/pull/54, part of https://github.com/onflow/cadence/issues/2412

This implements type checking for the creation and usage of authorized/entitled references, as well as subtyping and access control 
______

- [ ] Targeted PR against `master` branch
- [X] Linked to Github issue with discussion and accepted design OR link to spec that describes this work
- [X] Code follows the [standards mentioned here](https://github.com/onflow/cadence/blob/master/CONTRIBUTING.md#styleguides)
- [ ] Updated relevant documentation 
- [X] Re-reviewed `Files changed` in the Github PR explorer
- [X] Added appropriate labels

### CODE REVIEW

File: runtime/sema/type.go
Comment (dsainati1) [inline runtime/sema/type.go:5454]: !

File: runtime/sema/access.go
Comment (SupunS) [inline runtime/sema/access.go:278]: Maybe it's worth caching the resulting `EntitlementSetAccess` related to the `domain` and `subdomain` (below) in the `EntitlementMapAccess`

File: runtime/sema/access.go
Comment (SupunS) [inline runtime/sema/access.go:296]: similarly, maybe also cache the `output` here

File: runtime/sema/check_composite_declaration.go
Comment (SupunS) [inline runtime/sema/check_composite_declaration.go:105]: inversing these two conditions may help reduce the number of nested levels.

File: encoding/json/encode.go
Comment (dsainati1) [reply encoding/json/encode.go:616]: I already commented on the CCF issue about it yeah.

File: runtime/convertTypes.go
Comment (dsainati1) [reply runtime/convertTypes.go:473]: Correct, `auth &T` will not parse, the `auth` modifier will require an entitlement set as an argument.

File: runtime/interpreter/interpreter.go
Comment (dsainati1) [reply runtime/interpreter/interpreter.go:783]: Yes, they are mostly for me to keep track of what remains to do in later PRs. The only remaining `TODO` is for auth account entitlements, which still needs a FLIP.

File: runtime/sema/type.go
Comment (SupunS) [inline runtime/sema/type.go:3903]: I guess this includes any default functions coming from interfaces? i.e: entitlement isn't directly used in the composite type.
Maybe good to add a test if there aren't any.

File: runtime/sema/check_composite_declaration.go
Comment (dsainati1) [reply runtime/sema/check_composite_declaration.go:105]: Looks like the second condition here is actually unnecessary, so I removed it.

File: runtime/sema/type.go
Comment (dsainati1) [reply runtime/sema/type.go:3903]: Good call; this wasn't actually working properly. Fixed it and added a test.

File: runtime/interpreter/statictype.go
Comment (turbolent) [inline runtime/interpreter/statictype.go:648]: Needs metering

File: runtime/interpreter/statictype.go
Comment (turbolent) [inline runtime/interpreter/statictype.go:830]:
```suggestion
			entitlements = append(entitlements, key.ID())
```

File: runtime/interpreter/statictype.go
Comment (turbolent) [inline runtime/interpreter/statictype.go:836]:
```suggestion
		return NewEntitlementMapAuthorization(memoryGauge, key.ID())
```

File: runtime/sema/access.go
Comment (turbolent) [inline runtime/sema/access.go:269]: Great comment 👌

File: runtime/sema/access.go
Comment (turbolent) [inline runtime/sema/access.go:294]:
```suggestion
	domain := make(map[*EntitlementType]struct{})
```

File: runtime/sema/access.go
Comment (turbolent) [inline runtime/sema/access.go:307]:
```suggestion
	codomain := make(map[*EntitlementType]struct{})
```

File: runtime/sema/access.go
Comment (turbolent) [inline runtime/sema/access.go:319]:
```suggestion
	image := e.images[entitlement]
	if image != nil {
		return image
```

File: runtime/sema/access.go
Comment (turbolent) [inline runtime/sema/access.go:342]:
```suggestion
		output := orderedmap.New[EntitlementOrderedSet](inputs.Entitlements.Len())
```

File: runtime/sema/runtime_type_constructors.go
Comment (turbolent) [inline runtime/sema/runtime_type_constructors.go:245]:
```suggestion
		DocString: `
Creates a run-time type representing a reference type of the given type. 

The first argument specifies the set of entitlements to which this reference is entitled,
with the second specifying whether the set represents a conjunction or a disjunction. 

Providing a nil or empty array for the first argument will result in an unauthorized reference.
`,
```

File: runtime/sema/type.go
Comment (turbolent) [inline runtime/sema/type.go:650]: Avoid allocating a new empty set

```suggestion
	return nil
```

File: runtime/sema/type.go
Comment (turbolent) [inline runtime/sema/type.go:3904]: Use what was checked

```suggestion
	supportedEntitlements := t.supportedEntitlements
	if supportedEntitlements != nil {
		return supportedEntitlements
```

File: runtime/sema/type.go
Comment (turbolent) [inline runtime/sema/type.go:4112]:
```suggestion
		attachmentEntitlementAccess := attachment.attachmentEntitlementAccess
		if attachmentEntitlementAccess != nil {
			access = (*attachmentEntitlementAccess).Codomain()
```

File: runtime/sema/type.go
Comment (turbolent) [inline runtime/sema/type.go:4474]:
```suggestion
	supportedEntitlements := t.supportedEntitlements
	if supportedEntitlements != nil {
		return supportedEntitlements
```

File: runtime/storage_test.go
Comment (turbolent) [inline runtime/storage_test.go:1399]:
```suggestion
			"Typed":   &cadence.ReferenceType{
				Authorization: cadence.UnauthorizedAccess,
				Type: cadence.IntType{},
			},
```

File: runtime/tests/checker/account_test.go
Comment (turbolent) [inline runtime/tests/checker/account_test.go:766]: Check type before casting with `require.IsType`, so test failure does not crash tests.

(For this test and the other tests in this file and other files)

File: runtime/tests/checker/account_test.go
Comment (turbolent) [inline runtime/tests/checker/account_test.go:811]: Same as above, check type before casting

File: runtime/tests/checker/attachments_test.go
Comment (turbolent) [inline runtime/tests/checker/attachments_test.go:1454]: Why are these tests now succeeding?

File: runtime/tests/checker/nft_test.go
Comment (turbolent) [reply runtime/tests/checker/nft_test.go:74]: Could you please create a follow-up issue for that?

File: runtime/tests/checker/reference_test.go
Comment (turbolent) [inline runtime/tests/checker/reference_test.go:124]:
```suggestion
          entitlement X
```

**Review by turbolent — COMMENTED** [review]
Great work!

File: runtime/interpreter/decode.go
Comment (dsainati1) [reply runtime/interpreter/decode.go:1466]: This is implemented in https://github.com/onflow/cadence/pull/2421

File: runtime/interpreter/statictype.go
Comment (dsainati1) [reply runtime/interpreter/statictype.go:497]: a zero-length string still meters a quantity of `1` raw string usage.

File: runtime/interpreter/statictype.go
Comment (SupunS) [inline runtime/interpreter/statictype.go:576]: Could re-use this constant memory usage (`common.NewConstantMemoryUsage(common.MemoryKindEntitlementMapStaticAccess)`)
Have similar pre-defined ones in `common/metering.go`

**Review by SupunS — APPROVED** [review]
LGTM!

File: runtime/sema/type.go
Comment (turbolent) [reply runtime/sema/type.go:5553]: Could you please delete the file? It won't be accurate anymore after the merge of this PR

File: runtime/sema/access.go
Comment (turbolent) [inline runtime/sema/access.go:299]: Here and elsewhere: The result of [`maps.Keys`](https://pkg.go.dev/golang.org/x/exp/maps#Keys) (and `maps.Values`) has indeterminate order, it should not be used. Build up a slice yourself in the loop above, and maybe refactor the resulting code in a function and use it elsewhere.

We should probably enforce that `maps.Keys/Values` is not used in the codebase. All I can find is [depguard](https://github.com/OpenPeeDeeP/depguard), which unfortunately forbids whole package imports.

File: runtime/sema/check_composite_declaration.go
Comment (turbolent) [inline runtime/sema/check_composite_declaration.go:101]:
```suggestion
			if memberAccess, ok := member.Access.(EntitlementSetAccess); ok {
```

File: runtime/sema/type.go
Comment (turbolent) [inline runtime/sema/type.go:192]: Use interface `ast.HasPosition` instead of concrete `ast.Range`, so the range does not have to be created unless needed.

```suggestion
	TypeIndexingElementType(indexingType Type, positioned ast.HasPosition) (Type, error)
```

File: runtime/sema/check_composite_declaration.go
Comment (turbolent) [inline runtime/sema/check_composite_declaration.go:98]:
```suggestion
		if attachmentAccess, ok := attachmentAccess.(EntitlementMapAccess); ok {
```

File: runtime/sema/check_composite_declaration.go
Comment (turbolent) [inline runtime/sema/check_composite_declaration.go:630]:
```suggestion
	if attachmentAccess, ok := attachmentAccess.(EntitlementMapAccess); ok {
```

File: runtime/sema/check_composite_declaration.go
Comment (turbolent) [inline runtime/sema/check_composite_declaration.go:1960]: Chopping down the argument list makes it easier to read

```suggestion
		functionType := checker.functionType(
			function.Purity, 
			functionAccess, 
			function.ParameterList, 
			function.ReturnTypeAnnotation,
		)
```

File: runtime/sema/check_composite_declaration.go
Comment (turbolent) [inline runtime/sema/check_composite_declaration.go:2369]: Not sure, but it looks like `fnRange` is never used

File: runtime/sema/type.go
Comment (dsainati1) [reply runtime/sema/type.go:192]: Problem with this is that this function doesn't have a `checker` argument, which is needed to create a range from a positioned value. Is it worth adding this additional argument for this improvement?

Comment (turbolent) [reply runtime/sema/type.go:192]: Ah, I see. In that case maybe just make the parameter a function `func() ast.Range`

**Review by turbolent — COMMENTED** [review]
Looks good, great work! 

I think we can merge this after the last couple things that are still open have been merged, or feel free to address them in a follow-up PR.

In the future it might be good to break PRs up more into smaller chunks, which makes them easier to review.

**Pull request merged**
